NSFS | NC | get-bucket-policy-status fails with internal error (TypeError: policy.statement is not iterable)

### Environment info
- NooBaa Version: **5.15.0**
- Platform: **RHEL8**

### Actual behavior
1. get-bucket-policy works, but get-bucket-policy-status fails with internal error:

```sh
$ aws s3api --endpoint http://localhost:80 get-bucket-policy --bucket bucket1                                                                              
{
    "Policy": "{\"Statement\":[{\"Sid\":\"AllowEveryoneReadOnlyAccess\",\"Effect\":\"Allow\",\"Principal\":\"*\",\"Action\":[\"s3:GetObject\",\"s3:ListBucket\"],\"Resource\":[\"arn:aws:s3:::bucket1\",\"arn:aws:s3:::bucket1/*\"]}]}"
}

$ aws s3api --endpoint http://localhost:80 get-bucket-policy-status --bucket bucket1                                                                       

An error occurred (InternalError) when calling the GetBucketPolicyStatus operation (reached max retries: 4): We encountered an internal error. Please try again.  
```

From the log:
```js
2024-03-07T15:50:13.684579+09:00 server node[1833210]: [nsfs/1833210]    [L0] core.sdk.bucketspace_fs:: BucketSpaceFS.read_bucket_sdk_info: bucket_config_path /gpfs/nsfs/buckets/bucket1.json
2024-03-07T15:50:13.685137+09:00 server node[1833210]: [nsfs/1833210]    [L0] core.sdk.bucketspace_fs:: BucketSpaceFS.read_bucket_sdk_info: bucket_config_path /gpfs/nsfs/buckets/bucket1.json
2024-03-07T15:50:13.687272+09:00 server node[1833210]: [nsfs/1833210]    [L1] core.sdk.object_sdk:: validate_non_nsfs_bucket:  
  { 
    name: SENSITIVE-5cfd9f24ccdd38e2, 
    email: SENSITIVE-5cfd9f24ccdd38e2, 
    creation_date: '2024-03-06T03:38:40.902Z', 
    access_keys: [ { access_key: SENSITIVE-2ca9fc5184acac2a, secret_key: SENSITIVE-1c41343d78549359 } ], 
    nsfs_account_config: { uid: 0, gid: 10, new_buckets_path: '/gpfs/bucket-root', fs_backend: 'GPFS' }, 
    allow_bucket_creation: true, 
    _id: '65e7e540e397b951176d4055' 
  } { fs_root_path: '', fs_backend: 'GPFS' }
2024-03-07T15:50:13.687881+09:00 server node[1833210]: [nsfs/1833210]    [L1] core.endpoint.s3.s3_rest:: S3 REQUEST GET /bucket1?policyStatus op get_bucket_policy_status request_id ltgvbupc-dgo0b9-iz1 
  { 
    host: 'localhost', 
    'accept-encoding': 'identity', 
    'user-agent': 'aws-cli/1.23.2 Python/3.6.8 Linux/4.18.0-477.27.1.el8_8.x86_64 botocore/1.23.46', 
    'x-amz-date': '20240307T065013Z', 
    'x-amz-content-sha256': '...', 
    authorization: 'AWS4-HMAC-SHA256 ...' 
  }
2024-03-07T15:50:13.688515+09:00 server node[1833210]: [nsfs/1833210]    [L0] core.sdk.bucketspace_fs:: BucketSpaceFS.get_bucket_policy: Bucket name bucket1
2024-03-07T15:50:13.689263+09:00 server node[1833210]: [nsfs/1833210]    [L0] core.sdk.bucketspace_fs:: BucketSpaceFS.get_bucket_policy: policy 
  {  
    Statement: [ { 
      Sid: 'AllowEveryoneReadOnlyAccess', 
      Effect: 'Allow', 
      Principal: '*', 
      Action: [ 's3:GetObject', 's3:ListBucket' ], 
      Resource: [ 'arn:aws:s3:::bucket1', 'arn:aws:s3:::bucket1/*' ] 
    } ]
  }
2024-03-07T15:50:13.689969+09:00 server node[1833210]: [nsfs/1833210] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR 
  <?xml version="1.0" encoding="UTF-8"?>
  <Error>
    <Code>InternalError</Code>
    <Message>We encountered an internal error. Please try again.</Message>
    <Resource>/bucket1?policyStatus</Resource>
    <RequestId>ltgvbupc-dgo0b9-iz1</RequestId>
  </Error> 
  GET /bucket1?policyStatus {
    "host":"localhost",
    "accept-encoding":"identity",
    "user-agent":"aws-cli/1.23.2 Python/3.6.8 Linux/4.18.0-477.27.1.el8_8.x86_64 botocore/1.23.46",
    "x-amz-date":"20240307T065013Z",
    "x-amz-content-sha256":"...",
    "authorization":"...
  } 
  TypeError: policy.statement is not iterable
    at _is_policy_public (/usr/local/noobaa-core/src/endpoint/s3/ops/s3_get_bucket_policy_status.js:19:36)
    at Object.get_bucket_policy_status [as handler] (/usr/local/noobaa-core/src/endpoint/s3/ops/s3_get_bucket_policy_status.js:12:23)
    at async handle_request (/usr/local/noobaa-core/src/endpoint/s3/s3_rest.js:150:19)
    at async Object.s3_rest [as handler] (/usr/local/noobaa-core/src/endpoint/s3/s3_rest.js:65:9)

```


### Expected behavior
1. The output should be like below.

{
    "PolicyStatus": {
        "IsPublic": false
    }
}

### Steps to reproduce
1. See above

### More information - Screenshots / Logs / Other output

--- a/src/endpoint/s3/ops/s3_get_bucket_policy_status.js
+++ b/src/endpoint/s3/ops/s3_get_bucket_policy_status.js
@@ -2,6 +2,7 @@
 'use strict';
 
 const S3Error = require('../s3_errors').S3Error;
+const _ = require('lodash');
 
 /**
  * https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketPolicyStatus.html
@@ -16,27 +17,19 @@ async function get_bucket_policy_status(req) {
 // TODO: implemented according to current implementation of authorize_request_policy. should update when authorize_request_policy changed
 // full public policy defintion: https://docs.aws.amazon.com/AmazonS3/latest/dev/access-control-block-public-access.html#access-control-block-public-access-policy-status
 function _is_policy_public(policy) {
-    for (const statement of policy.statement) {
-        let principal_wildcard = false;
-        let resource_wildcard = false;
-        if (statement.effect === 'deny') {
-            return false;
-        }
-        for (const principal of statement.principal) {
-            if (principal.unwrap() === '*') {
-                principal_wildcard = true;
+    for (const statement of policy.Statement) {
+        if (statement.Effect === 'Allow' && statement.Principal) {
+            const statement_principal = statement.Principal.AWS ? statement.Principal.AWS : statement.Principal;
+            //although redundant, its technicly possible to have both wildcard and specific principal. 
+            //in this case the wildcard principal override the specific one
+            for (const principal of _.flatten([statement_principal])) {
+                if (principal.unwrap() === '*') {
+                    return true;
+                }
             }
-        }
-        for (const resource of statement.resource) {
-            if ((/[?*]/).test(resource)) {
-                resource_wildcard = true;
-            }
-        }
-        if (!principal_wildcard || !resource_wildcard) {
-            return false;
         }
     }
-    return true;
+    return false;
 }
 
 module.exports = {

--- a/src/test/unit_tests/test_s3_bucket_policy.js
+++ b/src/test/unit_tests/test_s3_bucket_policy.js
@@ -106,7 +106,7 @@ async function setup() {
     });
 }
 
-/*eslint max-lines-per-function: ["error", 1100]*/
+/*eslint max-lines-per-function: ["error", 1300]*/
 mocha.describe('s3_bucket_policy', function() {
     mocha.before(setup);
     mocha.it('should fail setting bucket policy when user doesn\'t exist', async function() {
@@ -1128,6 +1128,121 @@ mocha.describe('s3_bucket_policy', function() {
                     throw err;
                 }
             }
+        });
+    });
+
+    mocha.describe('get-bucket-policy status should work', async function() {
+
+        mocha.it('get-bucket-policy status should return true for public policy', async function() {
+            const self = this; // eslint-disable-line no-invalid-this
+            self.timeout(15000);
+            const s3_policy = {
+                Version: '2012-10-17',
+                //Effect allow and principal equals '*'. public policy
+                Statement: [
+                    {
+                        Action: ['s3:PutObject'],
+                        Effect: 'Allow',
+                        Principal: { AWS: "*" },
+                        Resource: [`arn:aws:s3:::${BKT}/*`],
+                    }
+                ]};
+            await s3_owner.putBucketPolicy({
+                Bucket: BKT,
+                Policy: JSON.stringify(s3_policy)
+            });
+            const res = await s3_owner.getBucketPolicyStatus({Bucket: BKT});
+            assert.strictEqual(res.PolicyStatus.IsPublic, true);
+        });
+
+        mocha.it('get-bucket-policy status should return false for non "*" principal', async function() {
+            const self = this; // eslint-disable-line no-invalid-this
+            self.timeout(15000);
+            const s3_policy = {
+                Version: '2012-10-17',
+                Statement: [
+                    {
+                        Action: ['s3:PutObject'],
+                        Effect: 'Allow',
+                        Principal: { AWS: user_a }, //principal is user
+                        Resource: [`arn:aws:s3:::${BKT}/*`],
+                    }
+                ]};
+            await s3_owner.putBucketPolicy({
+                Bucket: BKT,
+                Policy: JSON.stringify(s3_policy)
+            });
+            const res = await s3_owner.getBucketPolicyStatus({Bucket: BKT});
+            assert.strictEqual(res.PolicyStatus.IsPublic, false);
+        });
+
+        mocha.it('get-bucket-policy status should return false for Deny actions', async function() {
+            const self = this; // eslint-disable-line no-invalid-this
+            self.timeout(15000);
+            const s3_policy = {
+                Version: '2012-10-17',
+                Statement: [
+                    {
+                        Action: ['s3:PutObject'],
+                        Effect: 'Deny',
+                        Principal: { AWS: "*" },
+                        Resource: [`arn:aws:s3:::${BKT}/*`],
+                    }
+                ]};
+            await s3_owner.putBucketPolicy({
+                Bucket: BKT,
+                Policy: JSON.stringify(s3_policy)
+            });
+            const res = await s3_owner.getBucketPolicyStatus({Bucket: BKT});
+            assert.strictEqual(res.PolicyStatus.IsPublic, false);
+        });
+
+        mocha.it('get-bucket-policy status should work for pricipal as a string', async function() {
+            const self = this; // eslint-disable-line no-invalid-this
+            self.timeout(15000);
+            const s3_policy = {
+                Version: '2012-10-17',
+                Statement: [
+                    {
+                        Action: ['s3:PutObject'],
+                        Effect: 'Allow',
+                        Principal: "*",
+                        Resource: [`arn:aws:s3:::${BKT}/*`],
+                    }
+                ]};
+            await s3_owner.putBucketPolicy({
+                Bucket: BKT,
+                Policy: JSON.stringify(s3_policy)
+            });
+            const res = await s3_owner.getBucketPolicyStatus({Bucket: BKT});
+            assert.strictEqual(res.PolicyStatus.IsPublic, true);
+        });
+
+        mocha.it('get-bucket-policy principal should return true if at least one statement is public with principal * ', async function() {
+            const self = this; // eslint-disable-line no-invalid-this
+            self.timeout(15000);
+            const s3_policy = {
+                Version: '2012-10-17',
+                Statement: [
+                    {
+                        Action: ['s3:PutObject'],
+                        Effect: 'Allow',
+                        Principal: "*",
+                        Resource: [`arn:aws:s3:::${BKT}/*`],
+                    },
+                    {
+                        Action: ['s3:GetObject'],
+                        Effect: 'Deny',
+                        Principal: "*",
+                        Resource: [`arn:aws:s3:::${BKT}/*`],
+                    }
+                ]};
+            await s3_owner.putBucketPolicy({
+                Bucket: BKT,
+                Policy: JSON.stringify(s3_policy)
+            });
+            const res = await s3_owner.getBucketPolicyStatus({Bucket: BKT});
+            assert.strictEqual(res.PolicyStatus.IsPublic, true);
         });
     });
 });

